### PR TITLE
Fix #214, revamp commitment struct

### LINKED ISSUE #177: Incorrect state root comparison after transition
After transition is performed with committed transactions comparison is applyied against __old__ stateRoot.
 https://github.com/thehubbleproject/RedditHubble/blob/59affd3e26ac8c2f9c9ff945ad3b376e9e3d1c68/contracts/rollup.sol#L367
https://github.com/thehubbleproject/RedditHubble/blob/59affd3e26ac8c2f9c9ff945ad3b376e9e3d1c68/contracts/rollup.sol#L387

How it should be like:

* Case 1, dispute targeted 0th commitment: we need to compare against final state root of previous batch.
* Case 2, dispute targeted nth commitment: then we must compare against (n-1)th state root of same batch.

With both cases we need to reveal two consecutive state roots commitements in order to detect invalid transition

### LINKED ISSUE #214: Redesign commit struct
### What's wrong

In the disputeBatch function, we need the previous state root and the current commitment to dispute. 

Since we have 2 types of commitment, ideally we should have only 2 types of disputeBatch functions for each commitment. The problem is, to get the previous state root we need the previous commitment for inclusion proof.

We need to redesign the commitment struct, so that no matter what the previous commitment looks like, we all have the same way to access the previous state root.

### Desired property

- No need for duplicated function
- Low overhead to add/remove a field from the struct
- Pass fewer data in a function
- Avoid passing unnecessary data to a function

### Proposed solution

- Remaining question: What to do with the commitType?

```solidity
struct Batch {
    bytes32 batchRoot;
    address committer;
    uint256 finalisesOn;
    bytes32 depositRoot;
    bool withdrawn;
}

// I don't think batchRoot is the right name because it's not the hash of the batch.
batchRoot = hash(accountRoot, commitmentsRoot)


// We use bodyRoot here so the body could be flexible
struct Commitment {
    bytes32 stateRoot;
    bytes32 bodyRoot;
}


// ideally tokenType and feeReceiver should be compressed into txs.
struct TransferBody {
    uint256 tokenType;
    uint256 feeReceiver;
    uint256[2] signature;
    bytes txs;
}

struct TransferCommitment {
    bytes32 stateRoot;
    TransferBody body;   
}

struct MassMigrationBody {
    uint256 targetSpokeID;
    bytes32 withdrawRoot;
    uint256 tokenID;
    uint256 amount;
    uint256[2] signature;
    bytes txs;
}

struct MassMigrationCommitment {
    bytes32 stateRoot;
    MassMigrationBody body;   
}
```

```solidity
// this is the internal function for the common part
function submitBatch(
    bytes[] memory commitments,
) internal {
    Types.Batch memory newBatch = Types.Batch({
        batchRoot: keccak256(
            abi.encodePack(
                accountRegistry.root(),
                merkleUtils.getMerkleRootFromLeaves(commitments)
            )
        ),
        committer: msg.sender,
        finalisesOn: block.number + governance.TIME_TO_FINALISE(),
        depositRoot: ZERO_BYTES32,
        withdrawn: false
    });
    batches.push(newBatch);
    logger.logNewBatch(
        newBatch.committer,
        submissions[submissions.length - 1].updatedRoot,
        batches.length - 1,
        batchType
    );
}

```

```solidity
// These are the external API exposed to coordinators, customized for different type of commmitment
function submitNormalBatch(
    Types.TransferCommitment[] calldata commitment
) external payable onlyCoordinator {
    bytes32[] memory commitments = new bytes32[](submissions.length);
    for (uint256 i = 0; i < submissions.length; i++) {
        commitments[i] = keccak256(
            abi.encodePack(
                commitment.stateRoot,
                // Should we do the whole body here, or hash of body here?
                commitment.body.toHash()
            )
        );
    }
    submitBatch(commitments);
}

function submitMMBatch(
    Types.MassMigrationCommitment[] calldata commitment
) external payable onlyCoordinator {
    bytes32[] memory commitments = new bytes32[](submissions.length);
    for (uint256 i = 0; i < submissions.length; i++) {
        commitments[i] = keccak256(
            abi.encodePack(
                commitment.stateRoot,
                commitment.body.toHash()
            )
        );
    }
    submitBatch(commitments);
}
```

```solidity

modifier validPreviousCommit(
    uint256 batchId,
    uint256 targetPathToCommit,
    Types.CommitmentInclusionProof memory previous
) {
        // If target is the first commmitment in the batch, previous commitment is in previous batch
    if (targetPathToCommit == 0) {
        require(
            checkInclusion(
                batches[_batch_id - 1].commitmentRoot,
                previous
            ),
            "Rollup: previous commitment is absent in previous batch"
        );
    } else {
        // previous commit in current batch 
        require(
            previous.pathToCommitment == targetPathToCommit - 1,
            "Rollup: previous commitment is not a previous leaf"
        );
        require(
            checkInclusion(batches[_batch_id].commitmentRoot, previous),
            "Rollup: previous commitment is absent in current batch"
        );
    }
    _;
}

function disputeBatch(
    uint256 _batch_id,
    Types.CommitmentInclusionProof memory previous,
    Types.TransferCommitmentInclusionProof memory target,
    Types.AccountMerkleProof[] memory accountProofs
) 
    public
    isDisputable(_batch_id)
    validPreviousCommit(_batch_id, target.pathToCommitment, previous)
{
    require(
        checkInclusion(batches[_batch_id].commitmentRoot, target),
        "Commitment not present in batch"
    );

    bytes32 updatedBalanceRoot;
    bool isDisputeValid;
    (updatedBalanceRoot, isDisputeValid) = rollupReddit.processCommit(
        previous.commitment.stateRoot,
        target.commitment,
        accountProofs
    );
    if (isDisputeValid || updatedBalanceRoot != target.commitment.stateRoot) {
        invalidBatchMarker = _batch_id;
        SlashAndRollback();
        return;
    }
}

function disputeMMBatch(
    uint256 _batch_id,
    Types.CommitmentInclusionProof memory previous,
    Types.MMCommitmentInclusionProof memory target,
    Types.AccountMerkleProof[] memory accountProofs
) 
    public
    isDisputable(_batch_id)
    validPreviousCommit(_batch_id, target.pathToCommitment, previous)
{
    require(
        checkInclusion(batches[_batch_id].commitmentRoot, target),
        "Commitment not present in batch"
    );

    bytes32 updatedBalanceRoot;
    bool isDisputeValid;
    (updatedBalanceRoot, isDisputeValid) = rollupReddit.processMassMigrationCommit(
        previous.commitment.stateRoot,
        target.commitment,
        accountProofs
    );
    if (isDisputeValid || updatedBalanceRoot != target.commitment.stateRoot) {
        invalidBatchMarker = _batch_id;
        SlashAndRollback();
        return;
    }
}
```

### PR DESCRIPTION
**Pull request by ChihChengLiang**
Also fixes #177 .

### CODE REVIEW

File: test/MassMigration.test.ts [hunk 1]
```diff
@@ -45,6 +44,19 @@ describe("Mass Migrations", async function() {
 
         stateTree.createAccount(Alice);
         stateTree.createAccount(Bob);
+        const accountRoot = await registry.root();
+        const initialCommitment = MassMigrationCommitment.new(
+            stateTree.root,
```
Comment (ChihChengLiang) [inline test/MassMigration.test.ts:49]: In the test setup, we submit this state root as ground truth state.

File: contracts/libs/Types.sol [hunk 2]
```diff
@@ -29,39 +29,102 @@ library Types {
         bytes32 depositRoot;
         bool withdrawn;
     }
+    struct Commitment {
+        bytes32 stateRoot;
+        bytes32 bodyRoot;
+    }
```
Comment (ChihChengLiang) [inline contracts/libs/Types.sol:35]: To dispute a batch, we need the previous correct state root. With this Commitment struct, we can just get the state root without worrying if the commitment is a transfer commitment or Mass migration commitment.

**Review by ChihChengLiang — COMMENTED** [review]
Commented on some details and gotchas

File: contracts/Rollup.sol [hunk 3]
```diff
@@ -307,52 +312,36 @@ contract Rollup is RollupHelpers {
...
-                    MMInfo[i].amount,
-                    MMInfo[i].withdrawRoot,
-                    MMInfo[i].targetSpokeID,
-                    aggregatedSignatures[i],
-                    Types.Usage.MassMigration
-                )
-            );
+    /// @dev We don't define it as `external` function because of https://github.com/ethereum/solidity/issues/3199
+    function submitTransferBatch(Types.TransferCommitment[] memory commitments)
+        public
```
[the remark this answers is not in the document]
Comment (vaibhavchellani) [reply contracts/Rollup.sol:319]: But this forces us to send commitments as `memory` and not `calldata` and that's undesirable, correct?

File: contracts/Rollup.sol [hunk 4]
```diff
@@ -307,52 +312,36 @@ contract Rollup is RollupHelpers {
...
+        payable
+        onlyCoordinator
+    {
+        bytes32[] memory leaves = new bytes32[](commitments.length);
+        Types.TransferCommitment[] memory commitmentsCopy = commitments;
+        bytes32 accountRoot = accountRegistry.root();
+        for (uint256 i = 0; i < commitments.length; i++) {
+            commitmentsCopy[i].body.accountRoot = accountRoot;
```
[the remark this answers is not in the document]
Comment (vaibhavchellani) [reply contracts/Rollup.sol:327]: Why do we create a `commitmentsCopy`? Cant we mutate `commitments` directly in the loop?

[the remark this answers is not in the document]
Comment (ChihChengLiang) [reply contracts/Rollup.sol:327]: Oh, nice catch. Yes we can. The function was marked as external function, so need a copy to mutate the commitment. But now we can just mutate without copy.

**Comment by ChihChengLiang** [comment]
Some short conclusion from the discussion with @vaibhavchellani. Here are the execution cost of submitTransferBatch, with  32 (tx per commit) x 32 (commit per batch) = 1024 txs

1. 459,425: The cost of our Ropsten [record](https://ropsten.etherscan.io/tx/0x01c83dbce6894360a56dc6810106f47cbc699522a9844e126afc30f51abc0c2e), code [here](https://github.com/thehubbleproject/RedditHubble/blob/92eb1536d28e81a71b50fd857cebfe6c86742f27/contracts/rollup.sol#L229-L234).
2. 663,021: The one we use struct as args, and label the function as "public"
3. 519,000: Use no struct in args, and label the function as "external". see [here](https://github.com/thehubbleproject/RedditHubble/compare/commit-struct...using-external?expand=1)

3 compared to 2, we don't pass duplicated accountRoots in 3.
3 compared to 1, we have more fields like tokenType(uint256), feeReceiver(uint256), fees (4 bytes to each tx).

**Comment by ChihChengLiang** [comment]
Got offline approval from @vaibhavchellani

**Pull request merged**